### DISCUSSION TITLE
Move Python requirement to 3.10?

### DISCUSSION THREAD

**Discussion by stratofax** [TrustTheVote-Project/BallotLab#79] [Ideas]
Right now, we're testing BallotMaker against Python 3.8, 3.9, and 3.10 on Mac, Windows, and Linux. I've noticed some problems with VSCode reading import statements in 3.8 that seem to go away in 3.10. Also, I'm considering the use of some 3.10 features.

Since BallotMaker will ultimately run on a server, is there any reason to support older versions of Python? We can even ship BallotMaker as a Docker container. 

Can we drop the requirements to support Python 3.8 & 3.9, and just limit support to 3.10?

**Comment by ion-oset** [comment]
I lean conservatively on supported versions. One cannot count on having the latest and greatest of anything available on servers. The larger the client base of an organization the more cautious it tends to be on dealing with upgrades. This is particularly true of government organizations with large numbers of constituents and pressures to contend with.

I've proposed this baseline in the past (on the June 7 call), and I still believe it's a solid one: be prepared to run on deployments that support the current LTS of either Ubuntu Server or RedHat Enterprise. That means Ubuntu Server (20.04, 22.04) and RHEL (7, 8), which in turn means **Python 3.8 - 3.10**.

Yes, a server will likely run inside a Docker container. That doesn't mean there won't be situations where someone needs to debug or test on a machine with older versions, or incorporate code more directly into an existing service on an older version.

What features do you need from 3.10? How would you rank them in terms of necessity vs nice-to-have?

**Reply by stratofax** [reply]
These are good reasons to continue to support 3.8 & 3.9. I'm going to continue to develop in 3.10, even though I'm not using any 3.10 features right now, until I can resolve that strange import issue that seems to be related to 3.8, but the official test suite on GitHub Actions will continue to test 3.8, 3.9, and 3.10 on Mac, Windows and Linux.

Sadly, the tests are all failing on my dev branch since I started making ballots. It has nothing to do with the Python version but instead there are issues with grabbing external files (PNGs) from the repo. That's a completely separate issue, and I'll see if I can get the GitHub Actions back to green next week.

**Reply by stratofax** [reply]
All tests are green on GitHub Actions on my fork. This means:

- I fixed the file access issue (there were some Windows-specific path issues that pathlib handled nicely)
- All the code is executing successfully in Python 3.8, 3.9 and 3.10

So I think the answer to this question is "no," especially if I can develop in 3.10 but still test automatically in 3.8 and 3.9 across platforms.

**Reply by ion-oset** [reply]
> So I think the answer to this question is "no," especially if I can develop in 3.10 but still test automatically in 3.8 and 3.9 across platforms.

Agreed.
